File: 0094-binary-tree-inorder-traversal/0094-binary-tree-inorder-traversal.py
class Solution(object):
    def inorderTraversal(self, root):
        def traverse(root, arr):
            if root:
                traverse(root.left, arr)
                arr.append(root.val)
                traverse(root.right, arr)

        arr = []
        traverse(root, arr)
        return arr

File: 0094-binary-tree-inorder-traversal/0094-binary-tree-inorder-traversal.cpp
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        TreeNode* curr = root;
        vector<int> res;

        while (curr){
            if (curr->left == NULL){
                res.push_back(curr->val);
                curr = curr->right;
            }else{
                TreeNode* pred = curr->left;

                while (pred->right != NULL && pred->right != curr){
                    pred = pred->right;
                }

                if (pred->right == NULL){
                    pred->right = curr;
                    curr = curr->left;
                }else{
                    pred->right = NULL;
                    res.push_back(curr->val);
                    curr = curr->right;
                }
            }
        }

        return res;
    }
};